src/main/java/org/javarosa/entities/EntityFormPostProcessor.java

```diff
@@ -0,0 +1,67 @@
+package org.javarosa.entities;
+
+import kotlin.Pair;
+import org.javarosa.core.model.FormDef;
+import org.javarosa.core.model.IDataReference;
+import org.javarosa.core.model.instance.FormInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.entities.internal.EntitiesAttachment;
+import org.javarosa.entities.internal.EntityFormParseAttachment;
+import org.javarosa.form.api.FormEntryModel;
+import org.javarosa.form.api.FormPostProcessor;
+import org.javarosa.model.xform.XPathReference;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+
+public class EntityFormPostProcessor implements FormPostProcessor {
+
+    private static final String ENTITIES_NAMESPACE = "http://www.opendatakit.org/xforms/entities";
+
+    @Override
+    public void processForm(FormEntryModel formEntryModel) {
+        FormDef formDef = formEntryModel.getForm();
+        FormInstance mainInstance = formDef.getMainInstance();
+
+        EntityFormParseAttachment entityFormParseAttachment = formDef.getParseAttachment(EntityFormParseAttachment.class);
+        List<Pair<XPathReference, String>> saveTos = entityFormParseAttachment.getSaveTos();
+
+        String dataset = getDatasetToCreateWith(mainInstance);
+        if (dataset != null) {
+            List<Pair<String, String>> fields = saveTos.stream().map(saveTo -> {
+                IDataReference reference = saveTo.getFirst();
+                String answer = mainInstance.resolveReference(reference).getValue().getDisplayText();
+                return new Pair<>(saveTo.getSecond(), answer);
+            }).collect(Collectors.toList());
+
+            formEntryModel.putAttachment(new EntitiesAttachment(asList(new Entity(dataset, fields))));
+        } else {
+            formEntryModel.putAttachment(new EntitiesAttachment(emptyList()));
+        }
+    }
+
+    @Nullable
+    private String getDatasetToCreateWith(FormInstance mainInstance) {
+        TreeElement root = mainInstance.getRoot();
+        TreeElement meta = root.getChild("meta");
+        if (meta != null) {
+            TreeElement entity = meta.getChild("entity");
+
+            if (entity != null && entity.getNamespace().equals(ENTITIES_NAMESPACE)) {
```

Comment:
I feel like we should probably be checking this namespace (or searching with it) for all these elements. I'm thinking it would be good to look at unit tests for `EntityFormPostProcessor` to drive that out, but not sure how worth it is given how lax JavaRosa tends to be around namespaces usually?